jobs,webui: define a new endpoint for registry status

spaskob opened the pull request:
A request to _status/job_registry/{node_id} will
return info about jobs currently running on
this node.

Touches #45016.

Release note (admin ui change): end point /_status/job_registry/{node_id}
will display status info about the jobs running on this node_id.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45030)
<!-- Reviewable:end -->

Commit:
jobs,webui: define a new endpoint for registry status

A request to _status/registry/{node_id} will
return info about jobs currently running on
this node.

Touches #45016.

Release note (admin ui change): end point /_status/registry/{node_id}
will display status info about the jobs running on this node_id.

dt commented on pkg/jobs/registry.go:
```diff
@@ -179,6 +179,19 @@ func (r *Registry) MetricsStruct() *Metrics {
 }
 
+// Jobs returns a slice of the ids of all jobs running on this node.
+func (r *Registry) Jobs() []int64 {
```
nit: I'd call this something like `CurrentlyRunningJobs` or something.

dt commented on pkg/server/serverpb/status.proto:
```diff
@@ -753,6 +754,19 @@ message StatementsResponse {
   string internal_app_name_prefix = 4;
 }
 
+message RegistryStatusRequest {
```
nit: I'd call this *Job*RegistryStatus...

dt commented on pkg/server/serverpb/status.proto:
```diff
@@ -903,5 +917,9 @@ service Status {
   }
+  rpc RegistryStatus(RegistryStatusRequest) returns (RegistryStatusResponse) {
+    option (google.api.http) = {
+      get : "/_status/registry/{node_id}"
```
ditto on path.

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45030)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/server/serverpb/status.proto, line 922 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45030#-M-uf46hW8TS0jZj2OSX-r1-922:-M-vZJzi1ZLdzejVhkVq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d088ee2c69e5973296484da405d9670c9d4b0a7e/pkg/server/serverpb/status.proto#L922)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

ditto on path.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spaskob commented:
bors r+

spaskob commented:
bors r+

dt commented on pkg/jobs/registry.go:
```diff
@@ -179,6 +179,19 @@ func (r *Registry) MetricsStruct() *Metrics {
 
+// CurrentlyRunningJobs returns a slice of the ids of all jobs running on this node.
+func (r *Registry) CurrentlyRunningJobs() []int64 {
+	jobs := make([]int64, len(r.mu.jobs))
```
oh, sorry, I missed this earlier, but I think this has a race since we read the len before we acquire the lock, so if it then immediately grows, the iteration below could out-of-bounds.

We could hold the lock while allocating, or I think just switch this to `make([]int64, 0, len(...` and then use `ids = append(ids, id)` pattern below rather than baking in the exact length assumption,

spaskob replied:
Good catch - fixed in https://github.com/cockroachdb/cockroach/pull/45094
